Development

viordash opened the pull request:
Финальный проект модуля — разработка хранилища книг

viordash committed:
docker: configure git user

viordash committed:
docker: Switch package-tools to apt-get to fix installation error

Commit:
update tooling

Commit:
update vscode tooling

Commit:
book: GenreFromString func implementation

Commit:
tests: genre formatting test

Commit:
book: add ctors && formatter

Commit:
book: implementation of book  comparators

Commit:
concepts: implementation

Commit:
heterogeneous_lookup: implementation

Commit:
tests: fix group naming

Commit:
BookDatabase: implementation

Commit:
statsistics: implement buildAuthorHistogramFlat && calculateGenreRatings

Commit:
remove extra

Commit:
statsistics: implement calculateAverageRating && sampleRandomBooks

Commit:
statsistics: implement getTopNBy

Commit:
statsistics: improvements

Commit:
filters: implement YearBetween && filterBooks

Commit:
filters: implement RatingAbove && GenreIs && all_of && any_of

Commit:
calculateGenreRatings: add print formatter

Commit:
buildAuthorHistogramFlat: add print formatter

Commit:
main: uncomment code

Commit:
tests: add test for BookDatabase TypeAliases

Commit:
check-list: constexpr Book ctor

Commit:
check-list: replace key type in AuthorContainer to std::string_view

Commit:
check-list: replace key type in AuthorContainer to std::string_view

Commit:
check-list: use span for getTopNBy

Commit:
undo non-code changes

Commit:
remove unused includes

Commit:
remove unused includes

Nadezhda94 commented on include/book_database.hpp:
```diff
@@ -1,24 +1,26 @@
-    // Ваш код здесь
-
-    using AuthorContainer = BookContainer /* Ваш код здесь */;
+    using AuthorContainer = std::unordered_set<std::string_view, TransparentStringHash, TransparentStringEqual>;
```
unordered set отлично подойдет для хранения уникального множества авторов. 👍 
Но сейчас в нем хранятся string_view, которые ссылаются на строки снаружи, а они могут быть разрушены. Тут как раз нужно хранить в контейнере обычные std::string.

viordash replied:
я умышленно везде по коду использовал string_view, так как в include/book.hpp есть такое замечание:
`// string_view для экономии памяти, чтобы ссылаться на оригинальную строку, хранящуюся в другом контейнере`
И поэтому предположил, что паттерн с внешним хранилищем заложен в структуру проекта. Что в принципе логично на больших обьемах.  
Мне нужно от вас подтверждение, если все равно нужно применить string, то тогда я отдельным коммитом удалю этот комментарий, чтобы он не вводил в заблуждение.

Nadezhda94 replied:
Класс Book хранит author как string_view. Тут все верно.
А вот AuthorContainer будет хранить std::string. Это делается для того, чтобы Book ссылался на строку, которая не будет уничтожена раньше, чем Book. Храним std::string для авторов в AuthorContainer, Book содержит string_view для соответствующей строки из AuthorContainer.

Иначе, если AuthorContainer будет хранить string_view, может случиться так, что объект строки, на которую мы ссылаемся, снаружи будет уничтожен, а ссылка на него останется в AuthorContainer и в Book. 
Сохраняем std::string в AuthorContainer, меняем значение book.author так, чтобы оно теперь ссылалось на эту строку - тем самым гарантируем, что string_view поле класса Book будет ссылаться на валидную строку все время жизни нашего контейнера.

viordash replied:
теперь я понял причину, спасибо. 
Для ясности я добавил коммит, https://github.com/viordash/YP-Sprint-3-BookDatabase/pull/1/commits/3727a63caf97efbbe406a0228348795f7864540b с коррекцией "в другом контейнере". Я вероятно создам и PR в оригинальный репозиторий, если автор имел ввиду тоже самое, то это поможет будущим студентам не тратить время на неправильное решение.

Nadezhda94 commented on include/book_database.hpp:
```diff
@@ -31,6 +33,37 @@ class BookDatabase {
+    }
+
+    void PushBack(const Book &book) {
+        books_.push_back(book);
```
book тоже будет ссылаться на автора снаружи. Тут хочется сделать так, чтобы book ссылался на строку из AuthorContainer.

viordash replied:
done

Nadezhda94 commented on include/book_database.hpp:
```diff
@@ -31,6 +33,37 @@ class BookDatabase {
+
+    template <typename... Args>
+    void EmplaceBack(Args &&...args) {
+        books_.emplace_back(std::forward<Args>(args)...);
```
std::forward 👍

Nadezhda94 commented on include/comparators.hpp:
```diff
@@ -4,6 +4,16 @@
 
-struct LessByAuthor {};
+struct LessByAuthor {
+    bool operator()(const Book &lhs, const Book &rhs) const { return lhs.author < rhs.author; }
```
👍

Nadezhda94 commented on include/concepts.hpp:
```diff
@@ -7,19 +7,19 @@
 
 template <typename C>
-concept BookComparator = true;
+concept BookComparator = std::predicate<C, const Book&, const Book&>;
```
👍

Nadezhda94 commented on include/concepts.hpp:
```diff
@@ -7,19 +7,19 @@
-template <typename T>
-concept BookContainerLike = true;
+template <typename C>
+concept BookContainerLike = std::ranges::input_range<C> && std::same_as<std::ranges::range_value_t<C>, Book>;
```
Тут еще можно проверить, что контейнер поддерживает те методы, которые используются в реализации класса BookDatabase.
Например, { C.size() } -> std::convertible_to<std::size_t>;

Nadezhda94 commented on include/filters.hpp:
```diff
@@ -1,11 +1,41 @@
+
+template <BookPredicate... Preds>
+auto all_of(Preds... preds) {
+    return [=](const Book &book) { return (preds(book) && ...); };
```
Можно сэкономить на копировании [... ps = std::move(preds)]

viordash replied:
done

Nadezhda94 commented on include/filters.hpp:
```diff
@@ -1,11 +1,41 @@
+}
+
+template <BookIterator I, BookSentinel<I> S, BookPredicate P>
+auto filterBooks(I first, S last, P pred) {
```
👍

Nadezhda94 commented on include/heterogeneous_lookup.hpp:
```diff
@@ -1,14 +1,25 @@
+struct TransparentStringEqual {
+    using is_transparent = void;
+
+    bool operator()(std::string_view lhs, std::string_view rhs) const { return lhs == rhs; }
```
Еще важно добавить перегрузки, которые будут сравнивать std::string_view и std::string. Этот класс используется в AuthorContainer. Т.к. AuthorContainer будет хранить обычные std::string, важно находить элемент и по string_view, не создавая каждый раз строку из string_view. Для этого добавляем дополнительные перегрузки.

viordash replied:
хм, в случае если AuthorContainer будет хранить std::string, точно ли нужно добавлять перегрузки? При сравнении ведь должно происходить преобразование std::string в std::string_view, а не наоборот. Преобразование std::string в std::string_view не дорогое, и им лучше пренебречь, чем раздувать код.

Nadezhda94 replied:
В целом да, проблема была бы, если бы компаратор сравнивал только std::string. Тогда были бы ненужные преобразования к std::string от string_view.

Nadezhda94 commented on include/heterogeneous_lookup.hpp:
```diff
@@ -1,14 +1,25 @@
+    using is_transparent = void;
 
-struct TransparentStringEqual {};
+    bool operator()(std::string_view lhs, std::string_view rhs) const { return lhs < rhs; }
```
Важно добавить перегрузки, которые будут сравнивать std::string_view и std::string

viordash replied:
та же причина оставить std::string_view, как и для TransparentStringEqual

Nadezhda94 commented on include/heterogeneous_lookup.hpp:
```diff
@@ -1,14 +1,25 @@
+struct TransparentStringHash {
+    using is_transparent = void;
+
+    size_t operator()(std::string_view sv) const { return std::hash<std::string_view>{}(sv); }
```
Аналогично тут.

viordash replied:
та же причина оставить std::string_view, как и для TransparentStringEqual

Nadezhda94 commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,78 @@
+auto buildAuthorHistogramFlat(const BookDatabase<T> &cont) {
+    std::flat_map<std::string_view, int, Comparator> histogram;
+    for (const auto &book : cont.GetBooks()) {
+        auto [iterator, inserted] = histogram.insert({book.author, 0});
```
👍

Nadezhda94 commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,78 @@
+
+template <BookContainerLike T>
+double calculateAverageRating(const BookDatabase<T> &cont) {
+    if (cont.empty()) {
```
👍

Nadezhda94 commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,78 @@
+        return 0.0;
+    }
+    auto total =
+        std::accumulate(cont.begin(), cont.end(), 0.0, [](auto sum, const auto &book) { return sum + book.rating; });
```
Если нет зависимости от порядка выполнения операций, лучше использовать std::transform_reduce. Будет работать чуть быстрее.

viordash replied:
done

Nadezhda94 commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,78 @@
+    n = std::min(n, books.size());
+    result.reserve(n);
+
+    std::sort(books.begin(), books.end(), comp);
```
Тут может быть эффективнее использовать std::partial_sort.

viordash replied:
хм, а не получиться же применить partial_sort. У вас в main указано:
` auto topBooks = getTopNBy(db, 3, comp::LessByRating{});`
По логике получается что, в отсортированном массиве сначала будут идти элементы с наименьшим рейтингом.
И чтобы получить наивысший рейтинг придется сортировать все элементы, т.е. толку от частичной сортировки не будет. 
Или же нужно применить обратный компаратор, но в этом случае придется менять код в main. Я предполагаю, что код в нем не подлежит изменению.

Nadezhda94 commented on tests/statsistics_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+
+    auto genre_ratings = calculateGenreRatings(book_db.begin(), book_db.end());
+    ASSERT_EQ(genre_ratings.size(), 3);
+    ASSERT_FLOAT_EQ(genre_ratings[Genre::SciFi], 4.0);
```
ASSERT_DOUBLE_EQ тут корректнее, т.к. в проекте все вещественные значения в функциях, которые вычисляют статистики, имеют тип double.

Commit:
tests: to more float comparison accuracy

Commit:
statistics: replacing std::accumulate with std::transform_reduce

Commit:
statistics: refactoring getTopNBy

Commit:
Уточнение: string_view ссылается на AuthorContainer

Корректируем расплывчатое "другой контейнер" в комментарии Book,
явно указывая BookDatabase::AuthorContainer как источник строк

Commit:
review remarks: a guarantee of the security of BookDatabase::authors' links

Commit:
tests: add tests for ownership of BookDatabase::authors

Commit:
review remarks: all_of/any_of optimization via moving
